drivers: nrf: pinctrl: refactor and allow SoC-specific hooks

masz-nordic opened the pull request:
Refactor `pinctrl_nrf` driver and allow SoC-specifc hooks to be added.

masz-nordic committed:
drivers: pinctrl: nrf: use HAL

Migrate to HAL functions in NRF pinctrl.

Signed-off-by: Marcin Szymczyk <marcin.szymczyk@nordicsemi.no>

masz-nordic committed:
drivers: pinctrl: nrf: refactor

* Move per-peripheral configuration to separate functions.
* Iterate over a table of functions instead of huge switch-case.
* Move `pinctrl_soc.h` contents from `soc` to driver include path.
* Other minor improvements.

Signed-off-by: Marcin Szymczyk <marcin.szymczyk@nordicsemi.no>

masz-nordic committed:
soc: nordic: nrf54h: add SoC-specific pinctrl code

GPD is necessary for nRF54H20 application and radio code.
Provide hooks and select Kconfigs to enable them.

Signed-off-by: Marcin Szymczyk <marcin.szymczyk@nordicsemi.no>

bjarki-andreasen commented on soc/nordic/nrf54h/pinctrl.c:
```diff
@@ -0,0 +1,61 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <nrf/gpd.h>
```
something seems wrong with the includes here, this file should include `zephyr/drivers/pinctrl/nrf_pinctrl.h` (a file which does not exist yet, but it should, the hooks are device driver specific, not soc specific)

bjarki-andreasen commented on soc/nordic/common/pinctrl_soc.h:
```diff
@@ -26,6 +27,18 @@ extern "C" {
 typedef uint32_t pinctrl_soc_pin_t;
 
+
+struct nrf_pinctrl_pin {
```
Please move this struct and the hook declarations to a device driver specific header in include/zephyr/drivers/pinctrl/

these syms are not soc specific, they are device driver specific :) The SoC "merely" implements them :)

masz-nordic replied:
Moved whole `pinctrl_soc.h` to `include`, to align with how other vendors.

bjarki-andreasen reviewed (changes requested):
I love the refactor, and agree with the goal of the PR, but the locations of some syms seem off to me :)

masz-nordic committed:
drivers: pinctrl: nrf: add option to specify SoC-specific hooks

Instead of cluttering the generic driver, move SoC-specific
code to separate files.

Signed-off-by: Marcin Szymczyk <marcin.szymczyk@nordicsemi.no>

bjarki-andreasen commented:
ping @anangl

anangl commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -73,415 +35,657 @@ static const nrf_gpio_pin_drive_t drive_modes[NRF_DRIVE_COUNT] = {
+			pin.pin_num = PSEL_DISCONNECTED;
+		}
+
+		/* Iterate over pinctrl functions array.
+		 * If pin is handled by a given function, it returns true.
+		 */
+		bool handled = false;
+		uint8_t fun_idx = 0;
+
+		while (!handled && fun_idx < ARRAY_SIZE(funs_table)) {
+			handled = funs_table[fun_idx](&pin, &cfg);
+			fun_idx++;
+		}
```
I don't really understand the reason for this change. It is not required for introduction of the SoC-specific hooks and iterating over this array of functions is much less efficient than that "huge switch-case" that was here before. In the worst case, all these functions must be called before a pin is configured for the least lucky peripheral (and this is repeated for all the pins configured for a given peripheral). What's the benefit?

masz-nordic replied:
This will allow using pinctrl functions provided by nrfx drivers.

masz-nordic replied:
For peripherals supported on driver level, nrfx would expose API compatible with `nrf_pinctrl_fun` that replaces custom implementation in pincontrol SHIM.

anangl replied:
Such refactoring can be done then, if it indeed turns out to be beneficial. I don't see the point in doing this now.

anangl commented on include/zephyr/drivers/pinctrl/pinctrl_nrf_common.h:
```diff
@@ -0,0 +1,259 @@
+#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)			       \
+	{DT_FOREACH_CHILD_VARGS(DT_PHANDLE(node_id, prop),		       \
+				DT_FOREACH_PROP_ELEM_VARGS, psels,	       \
+				Z_PINCTRL_STATE_PIN_INIT, node_id)}
```
Why is all this stuff from `<pinctrl_soc.h>` moved here? In my opinion, we should have here only the hook declarations, `struct nrf_pinctrl_config`, and `NRF_GET_*` helpers (I don't see the need to introduce `struct nrf_pinctrl_pin`, hooks can be provided with the `pincfg` and `reg` values instead).

masz-nordic replied:
To keep everything in one place. This is in line with how other vendors do it:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/pinctrl/pinctrl_rcar_common.h#L100

anangl commented on soc/nordic/nrf54h/pinctrl.c:
```diff
@@ -0,0 +1,61 @@
+	return 0;
+}
+
+bool nrf_pinctrl_hook_custom_periph(const struct nrf_pinctrl_pin *pin,
+				    struct nrf_pinctrl_config *cfg)
+{
+	return false;
+}
```
How about having separate Kconfig options for particular hooks, so that such things could be avoided?

masz-nordic replied:
No strong feelings about this - for me it's just more readable to add empty C functions than more Kconfigs.

anangl commented on include/zephyr/drivers/pinctrl/pinctrl_nrf_common.h:
```diff
@@ -0,0 +1,259 @@
+ * @brief Hook called before GPIO properties configuration is written to register.
+ *
+ * @param pin Pointer to a structure describing pin that is currently handled.
+ * @param cfg Pointer to a structure describing current pin's configuration.
```
It's not the current configuration, but the configuration to be applied. And as I understand it, the hook is allowed to alter it, right?

masz-nordic replied:
True, I'll rephrase to `currently evaluated pin's configuration that will be applied`.

anangl commented on include/zephyr/drivers/pinctrl/pinctrl_nrf_common.h:
```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
```
2025
